File: 09_Heap_Priority_Queues/02_Last_Stone_Weight/1046-last-stone-weight.py
"""
Problem: LeetCode 1046 - Last Stone Weight

Key Idea:
To simulate the process of smashing stones, we can use a max-heap (priority queue) to keep track of the stone weights. At each step, we pop the two largest stones from the heap, smash them together, and push the resulting weight back into the heap. We repeat this process until there is only one stone left in the heap, which will be the last stone weight.

Time Complexity:
- Building the max-heap: The time complexity of building the max-heap is O(n), where n is the number of stones.
- Popping and pushing elements: The time complexity of each pop and push operation is O(log n), where n is the number of stones. We perform these operations until there is only one stone left.

Space Complexity:
- The space complexity is O(n), where n is the number of stones, due to the space required to store the max-heap.
"""

import heapq


class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        max_heap = [-stone for stone in stones]  # Use negative values for max-heap

        heapq.heapify(max_heap)

        while len(max_heap) > 1:
            x = -heapq.heappop(max_heap)  # Extract the largest stone
            y = -heapq.heappop(max_heap)  # Extract the second largest stone

            if x != y:
                heapq.heappush(max_heap, -(x - y))  # Push the remaining weight

        return -max_heap[0] if max_heap else 0

File: 09_Heap_Priority_Queues/02_Last_Stone_Weight/1046-last-stone-weight.rs
/*
Problem: LeetCode 1046 - Last Stone Weight

Key Idea:
The key idea is to use a max-heap (priority queue) to efficiently simulate the process of smashing stones until only one stone remains or no stones are left.

Approach:
1. Create a max-heap (priority queue) to store the weights of the stones.
2. Insert all stone weights into the max-heap.
3. While there are at least two stones in the max-heap:
   - Pop the two largest stones from the max-heap.
   - Calculate their difference (smash them together).
   - If the difference is not zero, insert it back into the max-heap.
4. If there is exactly one stone left in the max-heap, return its weight; otherwise, return 0.

Time Complexity:
- Heap Insertion and Extraction: O(log n) per operation, where n is the number of stones.
- In the worst case, we perform n-1 iterations of the while loop, resulting in O(n log n) time complexity for the entire process.

Space Complexity:
O(n), where n is the number of stones, as we use a max-heap to store the stone weights.
*/

use std::collections::BinaryHeap;

impl Solution {
    pub fn last_stone_weight(stones: Vec<i32>) -> i32 {
        let mut max_heap = BinaryHeap::new();

        // Insert all stone weights into the max-heap
        for stone in stones {
            max_heap.push(stone);
        }

        // Simulate stone smashing until only one stone remains or none
        while max_heap.len() >= 2 {
            let stone1 = max_heap.pop().unwrap();
            let stone2 = max_heap.pop().unwrap();
            let diff = stone1 - stone2;

            if diff > 0 {
                max_heap.push(diff);
            }
        }

        // Return the remaining stone weight or 0
        max_heap.pop().unwrap_or(0)
    }
}
